File: CodeForces/scanners/scanner.cpp
#include <iostream>
#include <bits/stdc++.h>
#include <string>
using namespace std;

class Scanner {
private:
    vector<string> buffer;

    void refill() {
        string line;
        do {
            if (!getline(cin, line)) return;
            size_t start = line.find_first_not_of(" \t\r\n");
            size_t end = line.find_last_not_of(" \t\r\n");
            if (start == string::npos) line.clear();
            else line = line.substr(start, end - start + 1);
        } while (line.empty());

        stringstream ss(line);
        string token;
        buffer.clear();
        while (ss >> token) buffer.push_back(token);
    }

public:
    int nextInt() {
        if (buffer.empty()) refill();
        int val = stoi(buffer.front());
        buffer.erase(buffer.begin());
        return val;
    }

    double nextDouble() {
        if (buffer.empty()) refill();
        double val = stod(buffer.front());
        buffer.erase(buffer.begin());
        return val;
    }

    long long nextLong() {
        if (buffer.empty()) refill();
        long long val = stoll(buffer.front());
        buffer.erase(buffer.begin());
        return val;
    }

    string nextWord() {
        if (buffer.empty()) refill();
        string val = buffer.front();
        buffer.erase(buffer.begin());
        return val;
    }

    string nextString() {
        if (buffer.empty()) refill();
        string val = buffer[0];
        for (size_t i = 1; i < buffer.size(); i++) {
            val += " " + buffer[i];
        }
        buffer.clear();
        return val;
    }

    vector<int> nextIntArray(bool sorted = false) {
        if (buffer.empty()) refill();
        vector<int> arr;
        arr.reserve(buffer.size());
        for (auto &s : buffer) arr.push_back(stoi(s));
        buffer.clear();
        if (sorted) sort(arr.begin(), arr.end());
        return arr;
    }

    vector<long long> nextLongArray(bool sorted = false) {
        if (buffer.empty()) refill();
        vector<long long> arr;
        arr.reserve(buffer.size());
        for (auto &s : buffer) arr.push_back(stoll(s));
        buffer.clear();
        if (sorted) sort(arr.begin(), arr.end());
        return arr;
    }
};

File: CodeForces/scanners/scanner.lua
Scanner = {}
Scanner.__index = Scanner

function Scanner:new()
    local obj = { buffer = {} }
    setmetatable(obj, self)
    return obj
end

-- Internal function to refill the buffer
function Scanner:refill()
    while #self.buffer == 0 do
        local line = io.read("*l")
        if line == nil then
            error("EOF reached")
        end
        for token in line:gmatch("%S+") do
            table.insert(self.buffer, token)
        end
    end
end

-- Public methods
function Scanner:nextInt()
    self:refill()
    local val = tonumber(table.remove(self.buffer, 1))
    if val == nil then error("Invalid integer token") end
    return math.floor(val)
end

function Scanner:nextDouble()
    self:refill()
    local val = tonumber(table.remove(self.buffer, 1))
    if val == nil then error("Invalid double token") end
    return val
end

function Scanner:nextLong()
    return self:nextInt()
end

function Scanner:nextWord()
    self:refill()
    local val = table.remove(self.buffer, 1)
    if val == nil then error("No word available") end
    return val
end

function Scanner:nextString()
    self:refill()
    local line = table.concat(self.buffer, " ")
    self.buffer = {}
    return line
end

function Scanner:nextIntArray(sorted)
    self:refill()
    local arr = {}
    for i = 1, #self.buffer do
        arr[i] = tonumber(self.buffer[i])
        if arr[i] == nil then error("Invalid integer in array") end
        arr[i] = math.floor(arr[i])
    end
    self.buffer = {}
    if sorted then table.sort(arr) end
    return arr
end

function Scanner:nextLongArray(sorted)
    return self:nextIntArray(sorted)
end
